### PR TITLE
Fix typing of ``_infer`` to allow a return value in the ``Generator``

### PR DESCRIPTION
**Pull request by DanielNoord**
## Steps

- [x] Write a good description on what the PR does.

## Description

Ref. https://github.com/PyCQA/astroid/pull/1653.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Related Issue

### CODE REVIEW

File: astroid/bases.py [hunk 1]
```diff
@@ -119,9 +118,9 @@ def __getattr__(self, name):
             return self.__dict__[name]
         return getattr(self._proxied, name)
 
-    def infer(
-        self: _ProxyT, context: InferenceContext | None = None, **kwargs: Any
-    ) -> collections.abc.Generator[_ProxyT, None, None]:
+    def infer(  # type: ignore[return]
```
Comment (DanielNoord) [inline astroid/bases.py:121]: `mypy` complains about wanting a `return None` statement whereas `pylint` complains about `useless-return`. Who is right here?

File: astroid/nodes/node_ng.py [hunk 2]
```diff
@@ -593,7 +593,7 @@ def _infer_name(self, frame, name):
 
     def _infer(
         self, context: InferenceContext | None = None, **kwargs: Any
-    ) -> Generator[InferenceResult, None, None]:
+    ) -> Generator[InferenceResult, None, InferenceErrorInfo | None]:
```
Comment (DanielNoord) [inline astroid/nodes/node_ng.py:596]: We already defined this for `FunctionDef._infer`

File: astroid/objects.py [hunk 3]
```diff
@@ -333,5 +333,5 @@ def infer_call_result(self, caller=None, context=None):
 
     def _infer(
         self: _T, context: InferenceContext | None = None, **kwargs: Any
-    ) -> Iterator[_T]:
+    ) -> Generator[_T, None, None]:
```
Comment (DanielNoord) [inline astroid/objects.py:336]: I thought I would fix this at the same time as well as it is somewhat related.

File: astroid/bases.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply astroid/bases.py:121]: I don't think you **need** an explicit return at the end of a generator, so I think mypy is wrong... But they say in their doc:

> You should give a statically typed function an explicit None return type even if it doesn’t return a value, as this lets mypy catch additional type errors:

So you must be explicit to help mypy it seems.

Comment (DanielNoord) [reply astroid/bases.py:121]: I think that refers to:
```python
def func(x: str):
    print(x)
```

Which should have a `-> None` as well.

**Review by cdce8p — APPROVED** [review]

File: astroid/bases.py (hunk 1 above)
Comment (DanielNoord) [reply astroid/bases.py:121]: Yeah I'll create a small reproducer in the coming days.

Comment (DanielNoord) [reply astroid/bases.py:121]: Reported at https://github.com/python/mypy/issues/13019.

Comment (DanielNoord) [reply astroid/bases.py:121]: @cdce8p See `mypy` issue. This is working as intended. Should we update `pylint` to allow `return None` for annotated functions?

Comment (cdce8p) [reply astroid/bases.py:121]: > But the real question is what should the default value be. mypy is used a lot so probably the one mypy chose to reduce friction.

I would agree. Both checks (mypy and pylint) are opinionated. The only thing for us is to make sure they don't collide if a users chooses to enable either one.

> Should we update `pylint` to allow `return None` for annotated functions?

Would it make sense to go one step further and disable `useless-return` entirely by default?

Comment (Pierre-Sassoulas) [reply astroid/bases.py:121]: > Would it make sense to go one step further and disable useless-return entirely by default?

Yeap, we don't need an option, my bad.

Comment (DanielNoord) [reply astroid/bases.py:121]: We should really tweak the default values of `pylint`.

Comment (Pierre-Sassoulas) [reply astroid/bases.py:121]: >We should really tweak the default values of pylint.

Agreed. I'm thinking about applying 3512 but there's a lot of other issues to handle and I never have the time. This is probably one of the big pain point of pylint right now though. The issue that take us a lot of time can be very niche in very specific cases sometime but default values are really freaking important. (Also invalid-name for < 3 char value. I get bitten by it for new project myself, first thing I need to do is always to disable this warning, someone trying pylint always talk to me about it, it's not a good look).

**Pull request merged**